### PR TITLE
ASoC: SOF: Intel: fix LNL DMA handling and enable chain_dma support

### PR DESCRIPTION
**Pull request by plbossart**
The DMA handling is not correct on capture when we have 4 channels aggregated from 2 DAIs on 2 different links. We currently setup the DMA for 2 channels, which causes the data to be recorded at half-speed.

Also this enables chain-DMA support for all interfaces, only tested on ALH playback for now.

Still draft since the capture w/ chain DMA does not work - xrun on capture.

### CODE REVIEW

**Comment by plbossart** [comment]
Moving to ready. This was tested in hardware, and the only problematic case is the capture which is also broken for MTL, so it's a firmware issue clearly.

**Comment by plbossart** [comment]
actually, need to fix checkpatch issues.

**Comment by plbossart** [comment]
wow, clang found an uninitialized variable....Amazing that other compilers don't detect this!

````
sound/soc/sof/intel/hda-dai.c:473:4: error: variable 'ch_mask' is uninitialized when used here [-Werror,-Wuninitialized]
                        ch_mask |= rtd->dai_link->codec_ch_maps[j].ch_mask;
                        ^~~~~~~
sound/soc/sof/intel/hda-dai.c:443:13: note: initialize the variable 'ch_mask' to silence this warning
        int ch_mask;
                   ^
                    = 0
````

**Review by bardliao — DISMISSED** [review]
LGTM

File: sound/soc/sof/intel/hda-dai.c [hunk 1]
```diff
@@ -452,9 +458,24 @@ int sdw_hda_dai_hw_params(struct snd_pcm_substream *substream,
...
-	/* in the case of SoundWire we need to program the PCMSyCM registers */
+	/*
+	 * in the case of SoundWire we need to program the PCMSyCM registers. In case
+	 * of aggregated devices, we need to define the channel mask for each sublink
+	 * by reconstructing the split done in soc-pcm.c
+	 */
+	for_each_rtd_cpu_dais(rtd, i, dai) {
+		if (dai == cpu_dai)
+			break;
+	}
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c:472]: Can we use other than `i`, like `cpu_dai_id`? It is not obvious right away that we only need the `i` form this loop later.

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 2]
```diff
@@ -237,15 +237,32 @@ static unsigned int generic_calc_stream_format(struct snd_sof_dev *sdev,
 					       struct snd_pcm_substream *substream,
 					       struct snd_pcm_hw_params *params)
 {
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai-ops.c:240]: typo in commit message:
capture introduces a **different** between the channels handled

and this reads a bit odd:
We need to reconstruct the number of channels to figure out how many channels, ...

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 3]
```diff
@@ -237,15 +237,32 @@ static unsigned int generic_calc_stream_format(struct snd_sof_dev *sdev,
...
+	 * if the multiple dais are handled by the same dailink, we may need to update the
+	 * stream channel count - the params are modified in soc-pcm based on the codec_ch_maps info
+	 */
+	for_each_rtd_codec_dais(rtd, j, codec_dai)
+		ch_mask |= rtd->dai_link->codec_ch_maps[j].ch_mask;
+
+	num_channels = hweight_long(ch_mask);
+	if (num_channels != params_channels(params))
+		dev_dbg(sdev->dev, "configuring stream format for %d channels, params_channels was %d\n",
+			num_channels, params_channels(params));
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai-ops.c:257]: do we need this print?

File: sound/soc/sof/ops.h [hunk 4]
```diff
@@ -555,6 +555,14 @@ snd_sof_set_mach_params(struct snd_soc_acpi_mach *mach,
 		sof_ops(sdev)->set_mach_params(mach, sdev);
 }
 
+static inline bool
+snd_sof_is_dai_supported(struct snd_sof_dev *sdev, u32 dai_type, bool dspless_mode)
```
Comment (ujfalusi) [inline sound/soc/sof/ops.h]: From the commit message it is not obvious why the is_dai_supported question is coupled with DSPless mode.

File: sound/soc/sof/intel/hda.c [hunk 5]
```diff
@@ -79,10 +78,26 @@ static u32 hda_get_interface_mask(struct snd_sof_dev *sdev)
...
 	return interface_mask[sdev->dspless_mode_selected];
 }
 
+bool hda_is_dai_supported(struct snd_sof_dev *sdev, u32 dai_type, bool dspless_mode)
+{
+	u32 interface_mask[2] = { 0 };
+
+	hda_get_interfaces(sdev, interface_mask);
+
+	return interface_mask[dspless_mode] & BIT(dai_type);
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.c]: I would be tempted to
```c
return !!(interface_mask[dspless_mode] & BIT(dai_type));
```

File: sound/soc/sof/intel/hda.c [hunk 6]
```diff
@@ -79,10 +78,26 @@ static u32 hda_get_interface_mask(struct snd_sof_dev *sdev)
...
+static u32 hda_get_interface_mask(struct snd_sof_dev *sdev)
+{
+	u32 interface_mask[2] = { 0 };
+
+	hda_get_interfaces(sdev, interface_mask);
 
 	return interface_mask[sdev->dspless_mode_selected];
 }
 
+bool hda_is_dai_supported(struct snd_sof_dev *sdev, u32 dai_type, bool dspless_mode)
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.c]: I see, ChainDMA can only be used on links which could be used in DSPless mode, both the host and link DMA is HD-DMA.

But is this a universal rule?

I'm not sure how this can be detached DSPless mode, perhaps `is_dai_feature_supported(sdev, dai_type, u8 feature)`
enum sof_dai_feature {
	sof_dai_feature_chain_dma,
};

if the dai_type is not supported, return false.
if the feature is not  supported by the dai type, return false.

I believe that there will be comment about the connection between ChainDMA and DSPless mode, but I'm not sure about the tying of the two.

Or just have a callback as blunt as: `is_chain_dma_supported(sdev, dai_type);`

File: sound/soc/sof/ipc4-topology.c [hunk 7]
```diff
@@ -530,10 +531,12 @@ static int sof_ipc4_widget_setup_comp_dai(struct snd_sof_widget *swidget)
 
 	pipe_widget = swidget->spipe->pipe_widget;
 	pipeline = pipe_widget->private;
-	if (pipeline->use_chain_dma && ipc4_copier->dai_type != SOF_DAI_INTEL_HDA) {
-		dev_err(scomp->dev,
-			"Bad DAI type '%d', Chained DMA is only supported by HDA DAIs (%d).\n",
-			ipc4_copier->dai_type, SOF_DAI_INTEL_HDA);
+
+	/* chain_dma is available for all DAIs on dspless platforms */
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: This is not true, TGL is 'dspless platform' but only HDA dais can support DSPless mode.

In comments we tend to use "DSPless"

**Review by ujfalusi — COMMENTED** [review]
@plbossart, I'm not sure about the tying of the ChainDMA to DSPless mode, is it future proof?

File: sound/soc/sof/ops.h [hunk 8]
```diff
@@ -555,6 +555,14 @@ snd_sof_set_mach_params(struct snd_soc_acpi_mach *mach,
...
+{
+	if (sof_ops(sdev) && sof_ops(sdev)->is_dai_supported)
+		return sof_ops(sdev)->is_dai_supported(sdev, dai_type, dspless_mode);
+	return false;
```
Comment (RanderWang) [inline sound/soc/sof/ops.h:564]: I prefer to have a blank line between two returns like other ops function in this file.

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 9]
```diff
@@ -237,15 +237,32 @@ static unsigned int generic_calc_stream_format(struct snd_sof_dev *sdev,
 					       struct snd_pcm_substream *substream,
 					       struct snd_pcm_hw_params *params)
 {
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *codec_dai;
 	unsigned int format_val;
+	u32 ch_mask = 0;
+	int num_channels;
+	int j;
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-dai-ops.c]: Why use j instead of i ?

**Review by RanderWang — DISMISSED** [review]
LGTM

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:472]: ok

Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:472]: done

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:240]: fixed and reworded completely, please recheck.

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c]: used codec_dai_id for symmetry with the cpu_dai loop in the first patch

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:257]: it's useful to visualize where the channel differences happen. I prefer to keep it, it helped me locate the configuration issue.

File: sound/soc/sof/intel/hda.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: done, see update.

File: sound/soc/sof/intel/hda.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: not sure why? Leaving as is for now.

File: sound/soc/sof/ipc4-topology.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c]: removed comment

File: sound/soc/sof/ops.h (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/ops.h]: fixed now

File: sound/soc/sof/ops.h (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/ops.h:564]: done

**Comment by plbossart** [comment]
@RanderWang @ujfalusi v2 updated, see diff 
[v2.diff.txt](https://github.com/thesofproject/linux/files/12590968/v2.diff.txt)

File: sound/soc/sof/intel/hda-dai.c [hunk 10]
```diff
@@ -452,9 +458,24 @@ int sdw_hda_dai_hw_params(struct snd_pcm_substream *substream,
...
 		return -ENODEV;
 
-	/* in the case of SoundWire we need to program the PCMSyCM registers */
+	/*
+	 * in the case of SoundWire we need to program the PCMSyCM registers. In case
+	 * of aggregated devices, we need to define the channel mask for each sublink
+	 * by reconstructing the split done in soc-pcm.c
+	 */
+	for_each_rtd_cpu_dais(rtd, cpu_dai_id, dai) {
+		if (dai == cpu_dai)
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-dai.c]: cpu_dai_id is not initialized and if cpu_dai can't match up with dai in this loop, cpu_dai_id is uncertain

Comment (lyakh) [reply sound/soc/sof/intel/hda-dai.c]: @RanderWang `cpu_dai_id` is used as a loop iterator
```
#define for_each_rtd_cpu_dais(rtd, i, dai)				\
	for ((i) = 0;							\
	     ((i) < rtd->dai_link->num_cpus) && ((dai) = asoc_rtd_to_cpu(rtd, i)); \
	     (i)++)
```
Maybe you meant, that if the `if` at line 467 never matches, the loop will run until the end, then `cpu_dai_id` will be equal to `rtd->dai_link->num_cpus` then the next loop won't find any matches either and so the channel mask will remain 0

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 11]
```diff
@@ -237,15 +237,32 @@ static unsigned int generic_calc_stream_format(struct snd_sof_dev *sdev,
...
 
-	format_val = snd_hdac_calc_stream_format(params_rate(params), params_channels(params),
+	/*
+	 * if the multiple dais are handled by the same dailink, we may need to update the
+	 * stream channel count - the params are modified in soc-pcm based on the codec_ch_maps info
+	 */
+	for_each_rtd_codec_dais(rtd, codec_dai_id, codec_dai)
+		ch_mask |= rtd->dai_link->codec_ch_maps[codec_dai_id].ch_mask;
+
+	num_channels = hweight_long(ch_mask);
```
Comment (lyakh) [inline sound/soc/sof/intel/hda-dai-ops.c:254]: this is more elegant, but counting inside the loop above might be faster and more compact

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-dai-ops.c:254]: I don't think that would be correct and even correct.
I'm not sure, but the different dais might use the same channel, so if you have 2 dais and both uses chan0/1 then the num_channels should end up to be 2.

Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:254]: no, we can't count inside the loop since some of the masks will overlap, so we might count the same channel multiple times. We have to check the mask after all the ORs are done.

File: sound/soc/sof/sof-priv.h [hunk 12]
```diff
@@ -157,6 +157,14 @@ struct sof_firmware {
 	u32 payload_offset;
 };
 
+enum sof_dai_feature {
+	SOF_DAI_FEATURE_DSP_ONLY, /* access from DSP only */
+	SOF_DAI_FEATURE_DSPLESS,  /* access from host only */
+	SOF_DAI_FEATURE_CHAIN_DMA,/* access from DSP with CHAIN_DMA mode */
```
Comment (ujfalusi) [inline sound/soc/sof/sof-priv.h]: I would re-word the comments and perhaps the SOF_DAI_FEATURE_DSP_ONLY is redundant?
I mean, one can ask if  a DAI can be used in DSPLESS mode and the answer is yes (it can be used in DSPless mode or with DSP mode) or no (it can only be used via DSP).
Similarly, the CHAIN_DMA question can be answered by: yes, ChainDMA can be used on the DAI or it cannot be used?

What would a HDA DAI report if you pass SOF_DAI_FEATURE_DSP_ONLY? It can run in DSPless mode, so it cannot say yes, but what no means? Can it support DSP mode?

File: sound/soc/sof/intel/hda-dai.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c]: @RanderWang has a point, there are no existing precedents of using the loop iterator outside the loop for this macro, and it could very well be wrong. Using a 'found' boolean would ensure that the default value of zero is not used. will fix.

File: sound/soc/sof/intel/hda.c [hunk 13]
```diff
@@ -45,44 +45,66 @@
...
 	case SOF_INTEL_CAVS_2_0:
 	case SOF_INTEL_CAVS_2_5:
 	case SOF_INTEL_ACE_1_0:
-		interface_mask[0] = BIT(SOF_DAI_INTEL_SSP) | BIT(SOF_DAI_INTEL_DMIC) |
-				    BIT(SOF_DAI_INTEL_HDA) | BIT(SOF_DAI_INTEL_ALH);
-		interface_mask[1] = BIT(SOF_DAI_INTEL_HDA);
+		interface_mask[SOF_DAI_FEATURE_DSP_ONLY] =
+			BIT(SOF_DAI_INTEL_SSP) | BIT(SOF_DAI_INTEL_DMIC) |
+			BIT(SOF_DAI_INTEL_HDA) | BIT(SOF_DAI_INTEL_ALH);
+		interface_mask[SOF_DAI_FEATURE_DSPLESS] = BIT(SOF_DAI_INTEL_HDA);
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda.c]: HDA is DSP_ONLY _and_ DSPLESSS(_ONLY) as well?

Comment (ujfalusi) [reply sound/soc/sof/intel/hda.c]: this is missing:
```c
interface_mask[SOF_DAI_FEATURE_CHAIN_DMA] = interface_mask[SOF_DAI_FEATURE_DSPLESS];
```
But likely this can be unconditional for Intel platforms?

**Review by ujfalusi — COMMENTED** [review]
I'm not sure about the definition of the `snd_sof_is_dai_supported` and especially the SOF_DAI_FEATURE_ definition.

Also, I think there is an issue with the dai validity check for Intel platforms.

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 11 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:254]: @ujfalusi each DAI has a channel mask, and we need to rebulld the dailink channel mask.

For amplifiers getting the same data, the dai masks are identical so the dailink mask will be the same as the dai masks.

For feedback capture, each dai has a different mask so the dailink will have a larger mask, the union of individual dai masks.

File: sound/soc/sof/sof-priv.h (hunk 12 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h]: maybe reword as DSP_ACCESS, HOST_ACCESS, DSP_ACCESS_CHAIN_DMA
The features are not exclusive or required, it's only the hardware restrictions that come into play to prevent one feature from being used on a specific platform

File: sound/soc/sof/intel/hda.c (hunk 13 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: Ah yes, for some reason I intentionally omitted it yesterday but this morning I can't recall what that reason was...

File: sound/soc/sof/sof-priv.h (hunk 12 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h]: @plbossart, I would expect all DAIs to be usable with via DSP, I find defining it redundant, I also understand why you want to have DSP_ACCESS in the enum - to be able to directly use it as an array index.
On the other hand these definitions beg to be a mask?
What about:
```
enum sof_dai_feature {
	SOF_DAI_DSP_ACCESS,,
	SOF_DAI_DSPLESS_ACCESS,
	SOF_DAI_CHAIN_DMA_SUPPORT,
	SOF_DAI_FEATURE_COUNT,
};

BIT(SOF_DAI_DSP_ACCESS)
BIT(SOF_DAI_DSPLESS_ACCESS) / BIT(SOF_DAI_HOST_ACCESS)
BIT(SOF_DAI_CHAIN_DMA_SUPPORT)
```
I would stick with the terms we are using for DSPLess and ChainDMA here.
The callback then can be 'check_dai_feature(sdev, dai, enum sof_dai_feature)`?

But does it make any sense at all? We have only one feature that we need the test for: ChainDMA
I think a single `dai_chain_dma_supported(sdev, dai)` or just a single feature definition to test for would be better and if there will be a need it an be extended, but I don't see that.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h]: well we have the existing interface_mask with the magic index 0 and 1, it'd be odd to have two ways of checking if an interface is supported in a specific more on a specific platform....

Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h]: @plbossart, what about dropping the DAI type check (in `sof_ipc4_widget_setup_comp_dai()`) for ChainDMA altogether and leave it to the topology writer to only use it on interfaces where it is allowed?
We fail that step currently if the DAI cannot be used with ChainDMA, if we drop that sanity check then we will likely fail a bit later in firmware I guess. Likely the error is going to be different, but if we have proper prints we will know the reason.

Comment (plbossart) [reply sound/soc/sof/sof-priv.h]: I don;t think it's a wise idea to trust the topology blindly. I'll be the first one to screw up...

Comment (ujfalusi) [reply sound/soc/sof/sof-priv.h]: That could be but the first time you try that tplg you will figure it out that it is not right and the final topology would be correct?
If we have the check for the DAI it will also fail, so it is fixed before the topology is ready for review.

I'm not sure if we do that or not, but we should print (dev_dbg) that a ChainDMA is used - but that is obvious form the messages we send, so likely no need to do that.

I just fear that we are trying to over engineer this for no benefits and the naming and definition of the check is not that clear. Especially the naming.

File: sound/soc/sof/intel/hda-dai.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c]: fixed.

File: sound/soc/sof/intel/hda.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: obsolete, moved to a non-ambiguous 'is_chain_dma_supported' callback

File: sound/soc/sof/sof-priv.h (hunk 12 above)
Comment (plbossart) [reply sound/soc/sof/sof-priv.h]: obsolete, moved to a non-ambiguous 'is_chain_dma_supported' callback

File: sound/soc/sof/intel/hda.c (hunk 13 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: obsolete, moved to a non-ambiguous 'is_chain_dma_supported' callback

**Comment by plbossart** [comment]
new version with simpler 'is_chain_dma_supported' callback and all comments addressed. Feedback welcome.

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/intel/hda-dai-ops.c [hunk 14]
```diff
@@ -637,16 +648,15 @@ hda_select_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swidg
 	case SOF_IPC_TYPE_4:
 	{
 		struct sof_ipc4_copier *ipc4_copier = sdai->private;
+		struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
 		const struct sof_intel_dsp_desc *chip;
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-dai-ops.c:653]: struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
struct sof_ipc4_copier *ipc4_copier = sdai->private;

**Review by RanderWang — DISMISSED** [review]
LGTM

File: sound/soc/sof/intel/hda-dai-ops.c (hunk 14 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai-ops.c:653]: done

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]
LGTM
